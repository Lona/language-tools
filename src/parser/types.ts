/* -------- Patterns -------- */

export type Identifiable = { id: string }

export type Extractable = { as?: string; label?: string }

export interface ConsumePattern<T> extends Identifiable, Extractable {
  type: 'consume'
  value: T
}

export interface SequencePattern<T> extends Identifiable, Extractable {
  type: 'sequence'
  value: Pattern<T>[]
  select?: string
}

export interface OrPattern<T> extends Identifiable, Extractable {
  type: 'or'
  value: Pattern<T>[]
  typeNames?: string[]
  parallel: boolean
}

export interface ManyPattern<T> extends Identifiable, Extractable {
  type: 'many'
  value: Pattern<T>
  separator?: Pattern<T>
}

export interface OptionPattern<T> extends Identifiable, Extractable {
  type: 'option'
  value: Pattern<T>
  select?: string
  defaultValue?: unknown
}

export interface PrintLiteralPattern<T> extends Identifiable {
  type: 'printLiteral'
  value: string
}

export type LineKind = 'soft' | 'hard'

export interface PrintLinePattern<T> extends Identifiable {
  type: 'printLine'
  kind?: 'soft' | 'hard'
}

export type TerminalPattern<T> = PrintLiteralPattern<T> | PrintLinePattern<T>

export interface PrintIndentPattern<T> extends Identifiable {
  type: 'printIndent'
  value: Pattern<T>
}

export interface PrintGroupPattern<T> extends Identifiable {
  type: 'printGroup'
  value: Pattern<T>
}

export interface PrintIfPattern<T> extends Identifiable {
  type: 'printIf'
  value: Pattern<T>
  reference: string
  before?: PrintLiteralPattern<T> | PrintLinePattern<T>
  after?: PrintLiteralPattern<T> | PrintLinePattern<T>
}

export interface ThunkPattern<T> extends Identifiable, Extractable {
  type: 'thunk'
  value: () => Pattern<T>
}

export type Pattern<T> =
  | ConsumePattern<T>
  | SequencePattern<T>
  | OrPattern<T>
  | ManyPattern<T>
  | OptionPattern<T>
  | ThunkPattern<T>
  // Printing
  | PrintLiteralPattern<T>
  | PrintLinePattern<T>
  | PrintIndentPattern<T>
  | PrintGroupPattern<T>
  | PrintIfPattern<T>

/* -------- Matches -------- */

export type ConsumeMatch<T> = {
  type: ConsumePattern<T>['type']
  match: string
}

export type SequenceMatch<T> = {
  type: SequencePattern<T>['type']
  match: string[]
}

export type OrMatch<T> = {
  type: OrPattern<T>['type']
  matchIndex: number
  match: string
}

export type ManyMatch<T> = {
  type: ManyPattern<T>['type']
  match: string[]
  matchSeparator: string[]
}

export type OptionMatch<T> = {
  type: OptionPattern<T>['type']
  match?: string
}

export type PrintLiteralMatch<T> = {
  type: PrintLiteralPattern<T>['type']
}

export type PrintLineMatch<T> = {
  type: PrintLinePattern<T>['type']
}

export type PrintIndentMatch<T> = {
  type: PrintIndentPattern<T>['type']
  match: string
}

export type PrintGroupMatch<T> = {
  type: PrintGroupPattern<T>['type']
  match: string
}

export type PrintIfMatch<T> = {
  type: PrintIfPattern<T>['type']
  match: string
}

export type ThunkMatch<T> = {
  type: ThunkPattern<T>['type']
  match: string
}

export type Match<T> =
  | ConsumeMatch<T>
  | SequenceMatch<T>
  | OrMatch<T>
  | ManyMatch<T>
  | OptionMatch<T>
  | ThunkMatch<T>
  // Printing
  | PrintLiteralMatch<T>
  | PrintLineMatch<T>
  | PrintIndentMatch<T>
  | PrintGroupMatch<T>
  | PrintIfMatch<T>

/* -------- Matches -------- */

export interface ConsumeParse<T, M> extends ConsumePattern<T> {
  match: M
}

export interface SequenceParse<T, M> extends SequencePattern<T> {
  match: Parse<T, M>[]
}

export interface OrParse<T, M> extends OrPattern<T> {
  matchIndex: number
  match: Parse<T, M>
}

export interface ManyParse<T, M> extends ManyPattern<T> {
  match: Parse<T, M>[]
  matchSeparator: Parse<T, M>[]
}

export interface OptionParse<T, M> extends OptionPattern<T> {
  match?: Parse<T, M>
}

export interface PrintLiteralParse<T, M> extends PrintLiteralPattern<T> {}

export interface PrintLineParse<T, M> extends PrintLinePattern<T> {}

export interface PrintIndentParse<T, M> extends PrintIndentPattern<T> {
  match: Parse<T, M>
}

export interface PrintGroupParse<T, M> extends PrintGroupPattern<T> {
  match: Parse<T, M>
}

export interface PrintIfParse<T, M> extends PrintIfPattern<T> {
  match: Parse<T, M>
}

export interface ThunkParse<T, M> extends ThunkPattern<T> {
  match: Parse<T, M>
}

export type Parse<T, M> =
  | ConsumeParse<T, M>
  | SequenceParse<T, M>
  | OrParse<T, M>
  | ManyParse<T, M>
  | OptionParse<T, M>
  | ThunkParse<T, M>
  // Printing
  | PrintLiteralParse<T, M>
  | PrintLineParse<T, M>
  | PrintIndentParse<T, M>
  | PrintGroupParse<T, M>
  | PrintIfParse<T, M>
