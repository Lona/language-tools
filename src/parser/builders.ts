import {
  ConsumePattern,
  SequencePattern,
  OrPattern,
  ManyPattern,
  OptionPattern,
  Pattern,
  PrintLiteralPattern,
  PrintLinePattern,
  PrintIndentPattern,
  PrintGroupPattern,
  ThunkPattern,
  PrintIfPattern,
  TerminalPattern,
  LineKind,
} from './types'
import { ParserPattern } from './Parser'

let initialId = 0

// Exported for testing
export function resetId() {
  initialId = 0
}

function id(): string {
  return String(initialId++)
}

type LabelOptions = { as?: string; label?: string }

type PatternShorthand<T> =
  | Pattern<T>
  | PatternShorthand<T>[]
  | (() => Pattern<T>)

function normalizePattern<T>(pattern: PatternShorthand<T>): Pattern<T> {
  if (typeof pattern === 'function') {
    return thunk(pattern)
  } else if (pattern instanceof Array) {
    return sequence(pattern)
  }

  return pattern
}

export function consume<T>(
  value: T,
  options: LabelOptions | string = {}
): ConsumePattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return { id: id(), type: 'consume', value, ...normalizedOptions }
}

export function sequence<T>(
  value: PatternShorthand<T>[],
  options?: { select?: string } & LabelOptions
): SequencePattern<T> {
  return {
    id: id(),
    type: 'sequence',
    value: value.map(normalizePattern),
    ...options,
  }
}

export function or<T>(
  value: PatternShorthand<T>[],
  options: LabelOptions & { typeNames?: string[]; parallel?: boolean } = {}
): OrPattern<T> {
  if (options.typeNames && value.length !== options.typeNames.length) {
    throw new Error('typeNames length must match number of patterns')
  }

  return {
    id: id(),
    type: 'or',
    value: value.map(normalizePattern),
    parallel: options.parallel ?? false,
    ...options,
  }
}

export function typedOr<T>(
  value: [string, PatternShorthand<T>][],
  options: LabelOptions & { parallel?: boolean } = {}
): OrPattern<T> {
  return {
    id: id(),
    type: 'or',
    value: value.map(([_, pattern]) => normalizePattern(pattern)),
    typeNames: value.map(([name]) => name),
    parallel: options.parallel ?? false,
    ...options,
  }
}

export function many<T>(
  value: PatternShorthand<T>,
  options: { separator?: PatternShorthand<T> } & LabelOptions = {}
): ManyPattern<T> {
  const { separator, ...rest } = options

  return {
    id: id(),
    type: 'many',
    value: normalizePattern(value),
    ...(separator && { separator: normalizePattern(separator) }),
    ...rest,
  }
}

export function option<T>(
  value: PatternShorthand<T>,
  options:
    | ({ select?: string; defaultValue?: unknown } & LabelOptions)
    | string = {}
): OptionPattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return {
    id: id(),
    type: 'option',
    value: normalizePattern(value),
    ...normalizedOptions,
  }
}

export function thunk<T>(
  value: () => Pattern<T>,
  options: LabelOptions | string = {}
): ThunkPattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return { id: id(), type: 'thunk', value, ...normalizedOptions }
}

// Printing

export function printLiteral<T>(value: string): PrintLiteralPattern<T> {
  return { id: id(), type: 'printLiteral', value }
}

export function printLine<T>(kind?: 'soft' | 'hard'): PrintLinePattern<T> {
  return { id: id(), type: 'printLine', ...(kind && { kind }) }
}

export function printIndent<T>(
  value: PatternShorthand<T>
): PrintIndentPattern<T> {
  return { id: id(), type: 'printIndent', value: normalizePattern(value) }
}

export function printGroup<T>(
  value: PatternShorthand<T>
): PrintGroupPattern<T> {
  return { id: id(), type: 'printGroup', value: normalizePattern(value) }
}

export function printIf<T>(
  value: PatternShorthand<T>,
  {
    before,
    after,
    reference,
  }: {
    reference: string
    before?: TerminalPattern<T>
    after?: TerminalPattern<T>
  }
): PrintIfPattern<T> {
  return {
    id: id(),
    type: 'printIf',
    value: normalizePattern(value),
    reference,
    ...(before && { before }),
    ...(after && { after }),
  }
}

export function printIndentedMany<T>(
  value: ManyPattern<T>,
  { padding }: { padding?: LineKind } = {}
) {
  if (!value.as) {
    throw new Error(`The 'printIndentedMany' helper requires an 'as' label`)
  }

  return printGroup(
    printIf(
      printIndent(
        printIf(value, { reference: value.as, before: printLine(padding) })
      ),
      { reference: value.as, after: printLine(padding) }
    )
  )
}

export function language(patterns: {
  [key: string]: PatternShorthand<string>
}): { [key: string]: ParserPattern } {
  return Object.fromEntries(
    Object.entries(patterns).map(([key, pattern]) => [
      key,
      normalizePattern(pattern),
    ])
  )
}
