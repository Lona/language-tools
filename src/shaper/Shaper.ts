import { ParserMatch, ParserPattern } from '../parser/Parser'
import { assert } from 'console'
import { inspect, isDeepStrictEqual } from 'util'
import { Parse } from '../parser/types'

type PatternType = string

export interface OrValue {
  type: string
  value: unknown
}

export interface ShaperOptions<Token, TokenShape, OrShape> {
  shapeToToken: (shape: TokenShape, type: PatternType) => Token
  tokenToShape: (token: Token) => TokenShape
  stringToShape: (string: string, type: PatternType) => TokenShape

  // Patterns
  orValueToShape: (value: OrValue, label?: string) => OrShape
  orShapeToValue: (shape: OrShape, label?: string) => OrValue
}

/**
 * Convert built-in types to a new "Shape"
 */
export class Shaper<Token, TokenShape, OrShape = OrValue> {
  options: ShaperOptions<Token, TokenShape, OrShape>

  constructor(options: ShaperOptions<Token, TokenShape, OrShape>) {
    this.options = options
  }

  extract = (match: Parse<string, Token>): unknown => {
    function wrapShape<T>(as: string, shape: T): unknown {
      return as === '.' ? shape : { [as]: shape }
    }

    // if ('as' in match) {
    //   console.log('as', match)
    // }

    switch (match.type) {
      case 'thunk':
        return wrapShape(match.as ?? '.', this.extract(match.match))
      case 'printGroup':
        return this.extract(match.match)
      case 'printIndent':
        return this.extract(match.match)
      case 'printLiteral':
        return null
      case 'printLine':
        return null
      case 'printIf':
        return this.extract(match.match)
      case 'consume':
        if (match.as) {
          const shape = this.options.tokenToShape(match.match)

          return wrapShape(match.as, shape)
        }

        return null
      case 'sequence': {
        const items = match.match.map(this.extract)
        const merged = Object.assign({}, ...items.filter((x) => x !== null))

        // Remove null fields
        const shape = Object.fromEntries(
          Object.entries(merged).flatMap(([key, value]) =>
            value !== null ? [[key, value]] : []
          )
        )

        if (match.select) {
          return (wrapShape(match.as ?? '.', shape) as any)[match.select]
        }

        return wrapShape(match.as ?? '.', shape)
      }
      case 'or':
        if (match.as) {
          const value = {
            type: String(
              match.typeNames
                ? match.typeNames[match.matchIndex]
                : match.matchIndex
            ),
            value: this.extract(match.match),
          }

          const shape = this.options.orValueToShape(value, match.label)

          return wrapShape(match.as, shape)
        }

        return this.extract(match.match)
      case 'many': {
        if (match.as) {
          const shape = match.match.map(this.extract)

          // TODO: Separator?
          return wrapShape(match.as, shape)
        }

        return match.match.map(this.extract)
      }
      case 'option': {
        if (match.as) {
          const defaultValue = match.defaultValue ?? null
          const shape = match.match ? this.extract(match.match) : defaultValue

          if (!isDeepStrictEqual(shape, defaultValue) && match.select) {
            return wrapShape(match.as, (shape as any)[match.select])
          }

          return wrapShape(match.as, shape)
        }

        return null
      }
    }
  }

  inject = (data: unknown, pattern: ParserPattern): Parse<string, Token> => {
    function unwrapValue<T>(
      as: string,
      container: T | { [key: string]: T }
    ): T {
      if (as === '.') return container as T

      assert(
        as in container,
        `expected '${as}' in container: ${inspect(container)}`
      )

      return (container as { [key: string]: T })[as]
    }

    function unwrapOptionalValue<T>(
      as: string,
      container: T | { [key: string]: T }
    ): T | null {
      if (as === '.') return container as T

      if (!(as in container)) return null

      return (container as { [key: string]: T })[as]
    }

    // console.log(
    //   [
    //     pattern.type,
    //     ...('label' in pattern ? [`(${pattern.label})`] : []),
    //     ...('as' in pattern ? [`'${pattern.as}'`] : []),
    //     inspect(data),
    //   ].join(' ')
    // )

    switch (pattern.type) {
      case 'thunk': {
        const value = unwrapValue<TokenShape>(pattern.as ?? '.', data as any)
        return { ...pattern, match: this.inject(value, pattern.value()) }
      }
      case 'printGroup':
      case 'printIndent':
      case 'printIf':
        return { ...pattern, match: this.inject(data, pattern.value) }
      case 'printLiteral':
        return { ...pattern }
      case 'printLine':
        return { ...pattern }
      case 'consume':
        if (pattern.as) {
          const value = unwrapValue<TokenShape>(pattern.as, data as any)

          return {
            ...pattern,
            match: this.options.shapeToToken(value, pattern.value),
          }
        } else {
          return {
            ...pattern,
            match: this.options.shapeToToken(
              this.options.stringToShape(pattern.value, pattern.value),
              pattern.value
            ),
          }
        }

      // throw new Error("Pattern 'consume' must have 'as' name")
      case 'or':
        if (pattern.as) {
          const orValue = unwrapValue<OrShape>(pattern.as, data as any)
          const orShape = this.options.orShapeToValue(orValue, pattern.label)
          const matchIndex = Number(
            pattern.typeNames
              ? pattern.typeNames.indexOf(orShape.type)
              : orShape.type
          )

          return {
            ...pattern,
            matchIndex,
            match: this.inject(orShape.value, pattern.value[matchIndex]),
          }
        }

        throw new Error("Pattern 'or' must have 'as' name")
      case 'sequence': {
        let sequenceValue = unwrapValue<unknown>(pattern.as ?? '.', data as any)

        if (pattern.select) {
          sequenceValue = { [pattern.select]: sequenceValue }
        }

        return {
          ...pattern,
          match: pattern.value.map((subpattern) =>
            this.inject(sequenceValue, subpattern)
          ),
        }

        // console.log(pattern)

        // throw new Error("Pattern 'sequence' must have 'as' name")
      }
      case 'many':
        if (pattern.as) {
          const manyValue = unwrapValue<unknown[]>(pattern.as, data as any)
          const separator = pattern.separator

          return {
            ...pattern,
            match: manyValue.map((item) => this.inject(item, pattern.value)),
            // Separators can't match data
            matchSeparator: separator
              ? manyValue.slice(0, -1).map(() => this.inject({}, separator))
              : [],
          }
        }

        throw new Error("Pattern 'many' must have 'as' name")
      case 'option': {
        if (pattern.as) {
          if (pattern.as === '.') {
            throw new Error("Option patterns can't use '.' for 'as' name")
          }

          let optionValue = unwrapOptionalValue<unknown>(
            pattern.as,
            data as any
          )
          const defaultValue = pattern.defaultValue ?? null

          if (isDeepStrictEqual(optionValue, defaultValue)) {
            return { ...pattern }
          }

          if (pattern.select) {
            optionValue = { [pattern.select]: optionValue }
          }

          return {
            ...pattern,
            match: this.inject(optionValue, pattern.value),
          }
        }

        throw new Error("Pattern 'option' must have 'as' name")
      }
    }
  }
}
