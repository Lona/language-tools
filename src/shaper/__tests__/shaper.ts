import {
  consume,
  many,
  option,
  or,
  printLine,
  sequence,
} from '../../parser/builders'
import { Parser } from '../../parser/Parser'
import { Shaper } from '../Shaper'

// Simple string tokens
let parser = new Parser({
  matchToken: (token: string, type: string) => token === type,
  tokenToValue: (token: string) => token,
})

const shaper = new Shaper<string, string>({
  stringToShape: (string: string) => string,
  shapeToToken: (string: string) => string,
  tokenToShape: (token: string) => token,
  orShapeToValue: (shape) => shape,
  orValueToShape: (value) => value,
})

it('shapes consume', () => {
  const pattern = consume('hello', { as: 'a' })
  const result = parser.parseOne(pattern, ['hello'])

  const shape = shaper.extract(result)
  expect(shape).toEqual({ a: 'hello' })

  const recreated = shaper.inject(shape, pattern)
  expect(recreated).toEqual(result)
})

describe('shapes or', () => {
  const pattern = or(
    [consume('hello', { as: 'b' }), consume('world', { as: 'c' })],
    {
      as: 'a',
    }
  )

  it('shapes case 0', () => {
    const result = parser.parseOne(pattern, ['hello'])

    const shape = shaper.extract(result)
    expect(shape).toMatchSnapshot()

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })

  it('shapes case 1', () => {
    const result = parser.parseOne(pattern, ['world'])

    const shape = shaper.extract(result)
    expect(shape).toMatchSnapshot()

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })
})

describe('shapes sequence', () => {
  const pattern = sequence(
    [consume('hello', { as: 'b' }), consume('world', { as: 'c' })],
    {
      as: 'a',
    }
  )

  const result = parser.parseOne(pattern, ['hello', 'world'])

  const shape = shaper.extract(result)
  expect(shape).toEqual({ a: { b: 'hello', c: 'world' } })

  const recreated = shaper.inject(shape, pattern)
  expect(recreated).toEqual(result)
})

describe('shapes many', () => {
  const pattern = many(consume('hello', { as: 'b' }), {
    as: 'a',
  })

  const result = parser.parseOne(pattern, ['hello', 'hello'])

  const shape = shaper.extract(result)
  expect(shape).toEqual({ a: [{ b: 'hello' }, { b: 'hello' }] })

  const recreated = shaper.inject(shape, pattern)
  expect(recreated).toEqual(result)
})

describe('shapes many with separator', () => {
  const pattern = many(consume('hello', { as: 'b' }), {
    separator: consume(','),
    as: 'a',
  })

  const result = parser.parseOne(pattern, ['hello', ',', 'hello', ',', 'hello'])

  const shape = shaper.extract(result)
  expect(shape).toEqual({ a: [{ b: 'hello' }, { b: 'hello' }, { b: 'hello' }] })

  const recreated = shaper.inject(shape, pattern)
  expect(recreated).toEqual(result)
})

describe('shapes many with line separator', () => {
  const pattern = many(consume('hello', { as: 'b' }), {
    separator: printLine(),
    as: 'a',
  })

  const result = parser.parseOne(pattern, ['hello', 'hello', 'hello'])

  const shape = shaper.extract(result)
  expect(shape).toEqual({ a: [{ b: 'hello' }, { b: 'hello' }, { b: 'hello' }] })

  const recreated = shaper.inject(shape, pattern)
  expect(recreated).toEqual(result)
})

describe('shapes option', () => {
  const pattern = option(consume('hello', { as: 'b' }), {
    as: 'a',
  })

  it('shapes existing option', () => {
    const result = parser.parseOne(pattern, ['hello'])

    const shape = shaper.extract(result)
    expect(shape).toEqual({ a: { b: 'hello' } })

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })

  it('shapes missing option', () => {
    const result = parser.parseOne(pattern, [])

    const shape = shaper.extract(result)
    expect(shape).toEqual({ a: null })

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })
})

describe('complex pattern', () => {
  const pattern = sequence(
    [
      consume('name', { as: 'name' }),
      option(
        sequence(
          [
            consume('<'),
            many(consume('Type', { as: 'type' }), {
              as: 'genericArguments',
              label: 'many type',
              separator: sequence([consume(','), printLine()]),
            }),
            consume('>'),
          ],
          {
            label: 'inner sequence',
          }
        ),
        {
          as: 'genericArguments',
          select: 'genericArguments',
          label: 'optional generics',
        }
      ),
    ],
    { label: 'complex sequence' }
  )

  it('nested optional without value', () => {
    const result = parser.parseOne(pattern, ['name'])

    const shape = shaper.extract(result)
    expect(shape).toMatchSnapshot()

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })

  it('nested optional with value', () => {
    const result = parser.parseOne(pattern, [
      'name',
      '<',
      'Type',
      ',',
      'Type',
      '>',
    ])

    const shape = shaper.extract(result)
    expect(shape).toMatchSnapshot()

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })
})
