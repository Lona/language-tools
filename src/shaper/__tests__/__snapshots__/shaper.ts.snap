// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex pattern nested optional with value 1`] = `
Object {
  "genericArguments": Array [
    Object {
      "type": "Type",
    },
    Object {
      "type": "Type",
    },
  ],
  "name": "name",
}
`;

exports[`complex pattern nested optional without value 1`] = `
Object {
  "name": "name",
}
`;

exports[`shapes or shapes case 0 1`] = `
Object {
  "a": Object {
    "type": "0",
    "value": Object {
      "b": "hello",
    },
  },
}
`;

exports[`shapes or shapes case 1 1`] = `
Object {
  "a": Object {
    "type": "1",
    "value": Object {
      "c": "world",
    },
  },
}
`;
